llvm/lib/Transforms/FuzzIntrospector/FuzzIntrospector.cpp

```diff
@@ -1132,3 +1184,234 @@ static RegisterStandardPasses
         PM.add(new FuzzIntrospector());
       });
 */
+
+void FuzzIntrospector::branchProfiler(Module &M) {
+
+  SmallPtrSet<const BasicBlock *, 32> visistedBBs;
+  SmallVector<const BasicBlock *, 32> worklist;
+  std::string outFileName = getNextLogFile() + ".branchProfile.yaml";
+  std::vector<BranchProfileEntry> outMap;
+
+
+  logPrintf(L1, "We are in branch profiler.\n");
+
+  for (const auto &F: M) {
+    // Skip declarations or the functions that are not wrapped e.g. not reachable from entry point
+    if(F.isDeclaration() || FuncComplexityMap.find(&F) == FuncComplexityMap.end()) {
+      continue;
+    }
+    auto fName = F.getName().str();
+    logPrintf(L1, "We are in branch profiler for %s\n", fName.c_str());
+
+    // This map is function level
+    std::map<BasicBlock *, size_t> BBComplexityMap;
+
+    for (const auto &BB: F) {
+      auto TI = BB.getTerminator();
+      auto BI = dyn_cast<BranchInst>(TI);
+      if (BI && BI->isConditional()) {
+        auto side0 = BI->getSuccessor(0); 
+        auto side1 = BI->getSuccessor(1);
+        auto BILoc = BI->getDebugLoc();
+
+        auto reachable0 = findReachables(side0);
+        auto reachable1 = findReachables(side1);
+
+        std::pair<size_t, size_t> complexities = findComplexities(reachable0, reachable1, BBComplexityMap);
```

Comment:
Please see below where I have some questions about `findComplexities`.

Reply:
This is now above :)

Reply:
Generally agreed, and addressed in prev comment :)